thunder/tests/test_inplace_functionalization.py

```diff
@@ -0,0 +1,124 @@
+    make = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    number = partial(make_number, dtype=dtype)
+
+    # pred_shape, a_shape, value
+    cases = (((2, 2, 2), (2, 2, 2), number()),)
+
+    for pred_shape, a_shape, value in cases:
+        pred, a = make(pred_shape, dtype=torch.bool, requires_grad=False), make(a_shape)
+        yield SampleInput(a, pred, value)
+
+
+_torchsymbol_to_torch: dict[Sybmol, Callable] = {v: k for k, v in _torch_to_thunder_function_map.items()}
+_functional_to_inplace: dict[Callable, Callable] = {
+    functional: inplace for inplace, (functional, index) in _inplace_to_out_of_place.items() if index == -1
+}
+_functional_to_functional_with_inplace_arg: dict[Callable, tuple[Callable, int]] = {
+    functional: (inplace, index) for inplace, (functional, index) in _inplace_to_out_of_place.items() if index >= 0
+}
+_inplace_opinfos: list[OpInfo] = []
+for op in opinfos:
+    if not (op.op in _functional_to_inplace or op.op in _functional_to_functional_with_inplace_arg):
+        continue
+    # ops that have an argument of `inplace` such as `F.relu` and `F.gelu`
+    if op.op in _functional_to_functional_with_inplace_arg:
+        inplace_op, _ = _functional_to_functional_with_inplace_arg[op.op]
+        assert op.name != "masked_fill"
+        inplace_opinfo = OpInfo(
+            inplace_op,
+            sample_input_generator=sample_generator_wrapper(op.sample_input_generator),
+            torch_reference=getattr(torch.nn.functional, op.name),
+        )
+        _inplace_opinfos.append(inplace_opinfo)
+    # in-place ops whose name ends with `_`
+    if op.op in _functional_to_inplace:
+        inplace_op = _functional_to_inplace[op.op]
+        inplace_opinfo = OpInfo(
+            inplace_op,
+            sample_input_generator=(
+                op.sample_input_generator if op.name != "masked_fill" else inplace_masked_fill_sample_generator
+            ),
+            torch_reference=_torchsymbol_to_torch[inplace_op],
+        )
+        _inplace_opinfos.append(inplace_opinfo)
+
+
+@dataclass(frozen=True)
+class InplaceOpWrapper:
+    torch_func: Callable
+    is_polygamma: bool
+    jitted: bool
+
+    def __call__(self, *args, **kwargs):
+        # polygamma expects an int as its first argument and a tensor as its second but
+        # torch.Tensor.polygamma_ wants opposite; tensor first, int second.
+        # ref:
+        # - https://pytorch.org/docs/stable/special.html#torch.special.polygamma
+        # - https://pytorch.org/docs/stable/generated/torch.Tensor.polygamma_.html
+        args = list(args)
+        idx = int(self.is_polygamma and self.jitted)
+        t = args[idx] + 1.0
+        args[idx] = t
+
+        self.torch_func(*args, **kwargs)
+        return t
+
+
+@ops(_inplace_opinfos, supported_dtypes=(dtypes.float32,))
+def test_functionalization(op: OpInfo, device: str, dtype: dtypes.dtype, executor, _):
+    import thunder
```

Comment:
Seems like we might be missing some test with a sequence of in-places ops?

Reply:
> Seems like we might be missing some test with a sequence of in-places ops?

yes, this pr doesn't have such tests.

> Just to be sure that the bsym replacement logic is sound across multiple symbols, not just a single one. Or is it not for this PR?

Locally I've been using the following snippet so I hope the functionalization works. I just didn't have a clear picture of designing tests with a sequence of in-place ops.

```python
import torch
import thunder


def f(a: torch.Tensor, b: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
    a += b
    c = torch.exp(a)
    d = torch.tanh(b)

    c += d

    d.div_(a)

    e = c + d
    f = torch.nn.functional.relu(e, inplace=True)
    g = a + b
    return f, c, d, torch.relu_(g)


def main():
    a, b = [torch.randn((2, 2), device="cuda", requires_grad=False) for _ in range(2)]
    a_, b_ = a.clone().detach(), b.clone().detach()
    jit_f = thunder.jit(f, executors=[thunder.executors.get_torch_executor()])
    c, d, e, g = jit_f(a, b)
    c_, d_, e_, g_ = f(a_, b_)

    traces = thunder.last_traces(jit_f)
    print(traces[-1])

    torch.testing.assert_close(d, d_)
    torch.testing.assert_close(c, c_)
    torch.testing.assert_close(e, e_)
    torch.testing.assert_close(g, g_)


if __name__ == "__main__":
    main()
```

Reply:
We could add some handpicked tests for sure. We can always throw some line duplication for numerically more stable ops :)